File: src-python/api/routes/facts.py
# Copyright (c) 2026 FicForge Contributors
# Licensed under the GNU Affero General Public License v3.0.
# See LICENSE file in the project root for full license text.

"""Facts 相关 API 路由。"""

from __future__ import annotations

from dataclasses import asdict
from pathlib import Path
from typing import Any, Optional

from fastapi import APIRouter, Query
from pydantic import BaseModel, Field
from starlette.concurrency import run_in_threadpool
from starlette.responses import JSONResponse

from api import (
    build_au_mutex,
    build_chapter_repository,
    build_fact_repository,
    build_ops_repository,
    build_project_repository,
    build_settings_repository,
    build_state_repository,
    error_response,
    validate_path,
)
from core.domain.enums import FactSource, FactStatus, FactType, NarrativeWeight
from core.services.facts_extraction import extract_facts_from_chapter, extract_facts_batch, load_character_aliases
from core.services.facts_lifecycle import FactsLifecycleError, add_fact, edit_fact, update_fact_status
from infra.llm.config_resolver import create_provider, resolve_llm_config

import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/facts", tags=["facts"])


class FactDataPayload(BaseModel):
    content_raw: str = ""
    content_clean: str = ""
    characters: list[str] = Field(default_factory=list)
    timeline: str = ""
    story_time: str = ""
    chapter: int | None = None
    status: FactStatus = FactStatus.ACTIVE
    type: FactType = FactType.PLOT_EVENT
    resolves: str | None = None
    narrative_weight: NarrativeWeight = NarrativeWeight.MEDIUM


class FactResponse(BaseModel):
    id: str
    content_raw: str
    content_clean: str
    characters: list[str]
    timeline: str
    story_time: str
    chapter: int
    status: FactStatus
    type: FactType
    resolves: str | None = None
    narrative_weight: NarrativeWeight
    source: FactSource
    revision: int
    created_at: str
    updated_at: str


class AddFactRequest(BaseModel):
    au_path: str
    chapter_num: int
    fact_data: FactDataPayload


class AddFactResponse(BaseModel):
    fact_id: str


class EditFactRequest(BaseModel):
    au_path: str
    updated_fields: dict[str, Any]


class EditFactResponse(BaseModel):
    fact_id: str
    revision: int


class UpdateFactStatusRequest(BaseModel):
    au_path: str
    new_status: FactStatus
    chapter_num: int


class UpdateFactStatusResponse(BaseModel):
    fact_id: str
    status: FactStatus


class ExtractFactsRequest(BaseModel):
    au_path: str
    chapter_num: int
    session_llm: Optional[dict[str, Any]] = None
    session_params: Optional[dict[str, Any]] = None


@router.get("", response_model=list[FactResponse])
async def list_facts(
    au_path: str = Query(...),
    status: FactStatus | None = Query(None),
    chapter: int | None = Query(None),
    characters: list[str] | None = Query(None),
) -> list[FactResponse] | JSONResponse:
    if not validate_path(au_path):
        return error_response(400, "INVALID_PATH", "路径不合法", [])
    repo = build_fact_repository()

    if characters:
        facts = await run_in_threadpool(repo.list_by_characters, au_path, characters)
    elif chapter is not None:
        facts = await run_in_threadpool(repo.list_by_chapter, au_path, chapter)
    elif status is not None:
        facts = await run_in_threadpool(repo.list_by_status, au_path, status)
    else:
        facts = await run_in_threadpool(repo.list_all, au_path)

    filtered = facts
    if status is not None and characters:
        filtered = [fact for fact in filtered if fact.status == status]
    if chapter is not None and characters:
        filtered = [fact for fact in filtered if fact.chapter == chapter]
    if status is not None and chapter is not None and not characters:
        filtered = [fact for fact in filtered if fact.status == status]

    return [FactResponse(**asdict(fact)) for fact in filtered]


@router.post("", response_model=AddFactResponse, status_code=201)
async def create_fact(request: AddFactRequest) -> AddFactResponse | JSONResponse:
    logger.info("Add fact: au=%s ch=%d", request.au_path, request.chapter_num)
    if not validate_path(request.au_path):
        return error_response(400, "INVALID_PATH", "路径不合法", [])
    if not (request.fact_data.content_clean or "").strip():
        return error_response(400, "INVALID_PARAMETER", "剧情笔记内容不能为空", [])
    repo = build_fact_repository()
    ops_repo = build_ops_repository()

    mutex = build_au_mutex()

    def _locked_add() -> Any:
        with mutex.get_lock(request.au_path):
            return add_fact(
                Path(request.au_path),
                request.chapter_num,
                request.fact_data.model_dump(exclude_none=True),
                repo,
                ops_repo,
            )

    try:
        fact = await run_in_threadpool(_locked_add)
    except FactsLifecycleError as exc:
        logger.exception("Add fact failed: au=%s ch=%d", request.au_path, request.chapter_num)
        return error_response(
            400,
            "ADD_FACT_INVALID",
            str(exc),
            ["检查 fact_data 字段是否合法"],
        )

    return AddFactResponse(fact_id=fact.id)


@router.put("/{fact_id}", response_model=EditFactResponse)
async def update_fact(fact_id: str, request: EditFactRequest) -> EditFactResponse | JSONResponse:
    logger.info("Edit fact: au=%s fact_id=%s", request.au_path, fact_id)
    if not validate_path(request.au_path):
        return error_response(400, "INVALID_PATH", "路径不合法", [])
    repo = build_fact_repository()
    ops_repo = build_ops_repository()
    state_repo = build_state_repository()

    mutex = build_au_mutex()

    def _locked_edit() -> Any:
        with mutex.get_lock(request.au_path):
            return edit_fact(
                Path(request.au_path),
                fact_id,
                request.updated_fields,
                repo,
                ops_repo,
                state_repo,
            )

    try:
        fact = await run_in_threadpool(_locked_edit)
    except FactsLifecycleError as exc:
        logger.exception("Edit fact failed: au=%s fact_id=%s", request.au_path, fact_id)
        return error_response(
            400,
            "EDIT_FACT_INVALID",
            str(exc),
            ["检查 fact_id 和 updated_fields"],
        )

    return EditFactResponse(fact_id=fact.id, revision=fact.revision)


@router.patch("/{fact_id}/status", response_model=UpdateFactStatusResponse)
async def patch_fact_status(fact_id: str, request: UpdateFactStatusRequest) -> UpdateFactStatusResponse | JSONResponse:
    logger.info("Update fact status: au=%s fact_id=%s status=%s", request.au_path, fact_id, request.new_status.value)
    if not validate_path(request.au_path):
        return error_response(400, "INVALID_PATH", "路径不合法", [])
    repo = build_fact_repository()
    ops_repo = build_ops_repository()
    state_repo = build_state_repository()

    mutex = build_au_mutex()

    def _locked_status() -> Any:
        with mutex.get_lock(request.au_path):
            return update_fact_status(
                Path(request.au_path),
                fact_id,
                request.new_status.value,
                request.chapter_num,
                repo,
                ops_repo,
                state_repo,
            )

    try:
        result = await run_in_threadpool(_locked_status)
    except FactsLifecycleError as exc:
        logger.exception("Update fact status failed: au=%s fact_id=%s", request.au_path, fact_id)
        return error_response(
            400,
            "UPDATE_FACT_STATUS_INVALID",
            str(exc),
            ["检查 fact_id、status 和 chapter_num"],
        )

    return UpdateFactStatusResponse(
        fact_id=result["fact_id"],
        status=FactStatus(result["new_status"]),
    )


class BatchStatusRequest(BaseModel):
    au_path: str
    fact_ids: list[str]
    new_status: FactStatus
    chapter_num: int = 0


class BatchStatusResponse(BaseModel):
    updated: int
    failed: int


@router.patch("/batch-status", response_model=BatchStatusResponse)
async def batch_update_status(request: BatchStatusRequest) -> BatchStatusResponse | JSONResponse:
    """批量更新 facts 状态。"""
    logger.info("Batch status: au=%s count=%d → %s", request.au_path, len(request.fact_ids), request.new_status.value)
    if not validate_path(request.au_path):
        return error_response(400, "INVALID_PATH", "路径不合法", [])

    repo = build_fact_repository()
    ops_repo = build_ops_repository()
    state_repo = build_state_repository()
    mutex = build_au_mutex()

    updated = 0
    failed = 0
    for fact_id in request.fact_ids:
        try:
            def _locked(fid: str = fact_id) -> Any:
                with mutex.get_lock(request.au_path):
                    return update_fact_status(
                        Path(request.au_path), fid, request.new_status.value,
                        request.chapter_num, repo, ops_repo, state_repo,
                    )
            await run_in_threadpool(_locked)
            updated += 1
        except Exception:
            failed += 1

    return BatchStatusResponse(updated=updated, failed=failed)


@router.post("/extract")
async def extract_facts_endpoint(request: ExtractFactsRequest) -> Any:
    """提取章节中的 facts 候选列表（PRD §6.7）。"""
    logger.info("Extract facts: au=%s ch=%d", request.au_path, request.chapter_num)
    if not validate_path(request.au_path):
        return error_response(400, "INVALID_PATH", "路径不合法", [])
    au_path = Path(request.au_path)
    chapter_repo = build_chapter_repository()
    fact_repo = build_fact_repository()
    project_repo = build_project_repository()
    settings_repo = build_settings_repository()

    try:
        content = await run_in_threadpool(
            chapter_repo.get_content_only, str(au_path), request.chapter_num,
        )
    except Exception:
        logger.exception("Extract facts chapter not found: au=%s ch=%d", request.au_path, request.chapter_num)
        return error_response(
            404,
            "CHAPTER_NOT_FOUND",
            f"章节 {request.chapter_num} 不存在",
            ["确认章节号是否正确"],
        )

    existing_facts = await run_in_threadpool(fact_repo.list_all, str(au_path))
    project = await run_in_threadpool(project_repo.get, str(au_path))
    settings = await run_in_threadpool(settings_repo.get)

    llm_config = resolve_llm_config(request.session_llm, project, settings)
    provider = create_provider(llm_config)

    # 从 project 中获取 cast_registry，从角色文件 frontmatter 读取别名
    cast_registry_obj = getattr(project, "cast_registry", None)
    cast_registry: dict[str, Any] = asdict(cast_registry_obj) if cast_registry_obj else {"characters": []}
    character_aliases = load_character_aliases(au_path)

    # 读取语言偏好
    _language = getattr(getattr(settings, "app", None), "language", "zh") or "zh"

    try:
        result = await run_in_threadpool(
            extract_facts_from_chapter,
            content,
            request.chapter_num,
            existing_facts,
            cast_registry,
            character_aliases,
            provider,
            llm_config,
            language=_language,
        )
    except Exception as exc:
        logger.exception("Extract facts failed: au=%s ch=%d", request.au_path, request.chapter_num)
        return error_response(
            500,
            "EXTRACT_FACTS_FAILED",
            f"提取失败: {exc}",
            ["检查 LLM 配置是否正确"],
        )

    return {"facts": [asdict(f) for f in result]}


# ---------------------------------------------------------------------------
# 批量提取（多章合并）
# ---------------------------------------------------------------------------

class ExtractFactsBatchRequest(BaseModel):
    au_path: str
    chapter_nums: list[int]
    session_llm: Optional[dict[str, Any]] = None


@router.post("/extract-batch")
async def extract_facts_batch_endpoint(request: ExtractFactsBatchRequest) -> Any:
    """批量提取多章 facts（合并为一个 LLM 调用）。"""
    logger.info("Extract facts batch: au=%s chapters=%s", request.au_path, request.chapter_nums)
    if not validate_path(request.au_path):
        return error_response(400, "INVALID_PATH", "路径不合法", [])
    au_path = Path(request.au_path)
    chapter_repo = build_chapter_repository()
    fact_repo = build_fact_repository()
    project_repo = build_project_repository()
    settings_repo = build_settings_repository()

    # 读取各章内容
    chapters_data: list[dict[str, Any]] = []
    for ch_num in request.chapter_nums:
        try:
            content = await run_in_threadpool(
                chapter_repo.get_content_only, str(au_path), ch_num,
            )
            chapters_data.append({"chapter_num": ch_num, "content": content})
        except Exception:
            logger.warning("Extract batch: chapter %d not found, skipping", ch_num)

    if not chapters_data:
        return error_response(404, "NO_CHAPTERS", "没有找到可提取的章节", [])

    existing_facts = await run_in_threadpool(fact_repo.list_all, str(au_path))
    project = await run_in_threadpool(project_repo.get, str(au_path))
    settings = await run_in_threadpool(settings_repo.get)

    llm_config = resolve_llm_config(request.session_llm, project, settings)
    provider = create_provider(llm_config)

    cast_registry_obj = getattr(project, "cast_registry", None)
    cast_registry: dict[str, Any] = asdict(cast_registry_obj) if cast_registry_obj else {"characters": []}
    character_aliases = load_character_aliases(au_path)

    # 读取语言偏好
    try:
        _language = getattr(getattr(settings, "app", None), "language", "zh") or "zh"
    except Exception:
        _language = "zh"

    try:
        result = await run_in_threadpool(
            extract_facts_batch,
            chapters_data,
            existing_facts,
            cast_registry,
            character_aliases,
            provider,
            llm_config,
            language=_language,
        )
    except Exception as exc:
        logger.exception("Extract facts batch failed: au=%s", request.au_path)
        return error_response(500, "EXTRACT_FACTS_FAILED", f"批量提取失败: {exc}", [])

    return {"facts": [asdict(f) for f in result]}

File: src-ui/src/api/facts.ts
// Copyright (c) 2026 FicForge Contributors
// Licensed under the GNU Affero General Public License v3.0.
// See LICENSE file in the project root for full license text.

/** Facts API */

import { FactType, NarrativeWeight } from "@ficforge/engine";

// 注：本文件不定义 FactInfo。UI 用的 FactInfo 真相源 = engine-client.ts 的
// `export type { Fact as FactInfo } from "@ficforge/engine"`（引擎完整 Fact，含全部富化字段）。
// 此处曾有一个 10 字段的窄版影子 interface（无人 import 的死类型），因其长得极像"该扩展的地方"
// 而屡次误导实施者，M3 批一删除（实施前调查·第一路发现附录①）。

export interface ExtractedFactCandidate {
  content_raw: string;
  content_clean: string;
  characters: string[];
  fact_type?: string;
  type?: string;
  narrative_weight: string;
  status: string;
  chapter: number;
  timeline?: string;
  // M8-A 富化 + 因果（提取已产出、运行时挂在候选对象上）。此前 UI 类型不声明 → 确认提取的
  // addFact 取不到 → 落库前丢（M9 spec §3.2 / 双审 BLOCKER）。声明出来才能转发。
  location?: string | null;
  story_time_tag?: string | null;
  story_time_order?: number | null;
  time_kind?: string | null;
  action_verb?: string | null;
  caused_by?: string[];
  known_to?: "all" | "reader_only" | string[] | null;
  hidden_from?: string[];
  suspense_type?: string | null;
  _confidence?: unknown; // 引擎侧 FactFieldConfidence；UI 不解释、只透传给 addFact，故用 unknown
  thread_ids?: string[]; // M8-B/M9：提取暂不产出，M9 自动挂线后经此转发（present 才带）
}

export interface ExtractFactsResponse {
  facts: ExtractedFactCandidate[];
}

/**
 * 「提取候选 → 事实入库 payload」的单一映射（盲审 2026-07-11 重复维：此前在
 * DirtyModal / useWriterFactsExtraction / useFactsExtraction / engine-chapters 四处
 * 手写并已实际漂移 —— DirtyModal 缺 content_raw 兜底、timeline 落空串）。
 * 兜底口径：content_raw 为空回退 content_clean；timeline 无值不带键。
 */
export function buildFactDataFromCandidate(c: ExtractedFactCandidate): Record<string, unknown> {
  return {
    content_raw: c.content_raw || c.content_clean,
    content_clean: c.content_clean,
    type: c.fact_type || c.type || FactType.PLOT_EVENT,
    narrative_weight: c.narrative_weight || NarrativeWeight.MEDIUM,
    status: c.status || "active",
    characters: c.characters || [],
    ...(c.timeline ? { timeline: c.timeline } : {}),
    ...extractedEnrichment(c), // caused_by + M8-A 富化
  };
}

/**
 * 从提取候选里抽出「确认落库时该一并带上」的富化/因果字段，供 addFact spread。
 * 单一真相源：三处确认路径（writer / FactsPage / DirtyModal）都用它，避免各自手维字段清单漂移。
 * 仅带「有值」的键（null/空数组/undefined 跳过），保持 addFact payload 干净。
 */
export function extractedEnrichment(c: ExtractedFactCandidate): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (c.location != null) out.location = c.location;
  if (c.story_time_tag != null) out.story_time_tag = c.story_time_tag;
  // known_to 用 != null 判据（与 hidden_from 的 length 判据不对称）是有意的：'all'/'reader_only'
  // 字符串态必须过；[] 空数组过了也无害 —— 引擎 addFact 的消毒器（domain/fact_sanitize）会把
  // [] 折叠为 null，落库口径单一，UI 不重复消毒（M3 批一）。
  if (typeof c.story_time_order === "number") out.story_time_order = c.story_time_order;
  if (c.time_kind != null) out.time_kind = c.time_kind;
  if (c.action_verb != null) out.action_verb = c.action_verb;
  if (c.caused_by?.length) out.caused_by = c.caused_by;
  if (c.known_to != null) out.known_to = c.known_to;
  if (c.hidden_from?.length) out.hidden_from = c.hidden_from;
  if (c.suspense_type != null) out.suspense_type = c.suspense_type;
  if (c._confidence != null) out._confidence = c._confidence;
  if (c.thread_ids?.length) out.thread_ids = c.thread_ids;
  return out;
}
